blockscout-ens/bens-logic/src/protocols/protocoler.rs

```diff
@@ -0,0 +1,229 @@
+use crate::{
+    blockscout::BlockscoutClient,
+    protocols::{DomainNameOnProtocol, ProtocolError},
+};
+use anyhow::anyhow;
+use ethers::{addressbook::Address, prelude::Bytes};
+use nonempty::NonEmpty;
+use sea_query::{Alias, IntoTableRef, TableRef};
+use serde::{Deserialize, Deserializer, Serialize};
+use std::{collections::HashMap, sync::Arc};
+
+#[derive(Debug, Clone)]
+pub struct Protocoler {
+    networks: HashMap<i64, Network>,
+    protocols: HashMap<String, Protocol>,
+}
+
+#[derive(Debug, Clone)]
+pub struct Network {
+    pub blockscout_client: Arc<BlockscoutClient>,
+    pub use_protocols: NonEmpty<String>,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Default)]
+pub struct Protocol {
+    pub info: ProtocolInfo,
+    pub subgraph_schema: String,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Default)]
+pub struct ProtocolInfo {
+    pub slug: String,
+    pub tld_list: NonEmpty<Tld>,
+    pub subgraph_name: String,
+    pub address_resolve_technique: AddressResolveTechnique,
+    pub empty_label_hash: Option<Bytes>,
+    pub native_token_contract: Option<Address>,
+    pub meta: ProtocolMeta,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Default)]
+pub struct ProtocolMeta {
+    pub short_name: String,
+    pub title: String,
+    pub description: String,
+    #[serde(default)]
+    pub icon_url: Option<String>,
+    #[serde(default)]
+    pub docs_url: Option<String>,
+}
+
+#[derive(Debug, Clone, Serialize, PartialEq, Eq, Default)]
+pub struct Tld(pub String);
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Default, Hash)]
+#[serde(rename_all = "snake_case")]
+pub enum AddressResolveTechnique {
+    #[default]
+    ReverseRegistry,
+    AllDomains,
+}
+
+impl Tld {
+    pub fn new(tld: &str) -> Tld {
+        Self(tld.trim_start_matches('.').to_string())
+    }
+
+    pub fn from_domain_name(name: &str) -> Option<Tld> {
+        name.rsplit_once('.').map(|(_, tld)| Self::new(tld))
+    }
+}
+
+impl<'de> Deserialize<'de> for Tld {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        String::deserialize(deserializer).map(|s| Self::new(&s))
+    }
+}
+
+impl Protocoler {
+    pub fn initialize(
+        networks: HashMap<i64, Network>,
+        protocols: HashMap<String, Protocol>,
+    ) -> Result<Self, anyhow::Error> {
+        networks
+            .values()
+            .map(|network| {
+                network
+                    .use_protocols
+                    .iter()
+                    .map(|name| {
+                        if protocols.contains_key(name) {
+                            Ok(())
+                        } else {
+                            Err(anyhow!("unknown protocol {name}"))
+                        }
+                    })
+                    .collect::<Result<Vec<_>, _>>()
+            })
+            .collect::<Result<Vec<_>, _>>()?;
```

Comment:
Nit: it may better convey the main idea of that code snippet
```suggestion
        for network in networks.values() {
            if let Some(name) = network
                .use_protocols
                .iter()
                .find(|&name| !protocols.contains_key(name))
            {
                return Err(anyhow!("unknown protocol {name}"));
            }
        }
```